File:  other-language-samples/factorial.ts
// const factorial = (n: number): number => {
//   if (n === 0) return 1;
//   if (n < 0) throw Error("negative number argument");

//   for (let i = n - 1; i > 1; i--) {
//     n *= i;
//   }
//   return n;
// };

const factorial = (n: number): number => {
  if (n === 0) return 1;
  if (n < 0) throw Error("negative number argument");
  return n * factorial(n - 1);
};

File: src/Factorial.hs
module Factorial
  ( factorial,
  )
where

factorial :: Integer -> Integer

-- パターンマッチとガードを使った再帰
-- factorial 0 = 1
-- factorial n
--   | n < 0 = error "negative argument"
--   | otherwise = n * factorial (n - 1)

-- 別解
factorial n
  | n < 0 = error "negative argument"
  | otherwise = product [1 .. n]
